src/error.rs

```diff
@@ -0,0 +1,170 @@
+use std::collections::BTreeMap;
+use std::num::{ParseFloatError, ParseIntError};
+use std::result;
+use std::string::FromUtf8Error;
+
+use base64::DecodeError;
+use quick_xml::Error as XmlError;
+use thiserror::Error as ThisError;
+
+use crate::Value;
+
+/// Errors that can occur when trying to perform an XML-RPC request.
+///
+/// This can be a lower-level error (for example, the HTTP request failed), a problem with the
+/// server (maybe it's not implementing XML-RPC correctly), or just a failure to execute the
+/// operation.
+#[derive(ThisError, Debug)]
+pub enum Error {
+    /// The response could not be parsed. This can happen when the server doesn't correctly
+    /// implement the XML-RPC spec.
+    #[error("parse error: {0}")]
+    ParseError(#[from] ParseError),
+
+    /// The response could not be encoded.
+    #[error("encoding error: {0}")]
+    EncodingError(#[from] EncodingError),
+
+    /// The server returned a `<fault>` response, indicating that the execution of the call
+    /// encountered a problem (for example, an invalid (number of) arguments was passed).
+    #[error("server fault: {} ({})", .0.fault_string, .0.fault_code)]
+    Fault(#[from] Fault),
+}
+
+#[derive(ThisError, Debug)]
+pub enum ParseError {
+    /// Error while parsing (malformed?) XML.
+    #[error("malformed XML: {0}")]
+    XmlError(#[from] XmlError),
+
+    #[error("malformed XML: {0}")]
+    ParseIntError(#[from] ParseIntError),
+
+    #[error("malformed XML: {0}")]
+    ParseFloatError(#[from] ParseFloatError),
+
+    #[error("malformed XML: {0}")]
+    Base64DecodeError(#[from] DecodeError),
+
+    #[error("malformed UTF-8: {0}")]
+    Utf8Error(#[from] FromUtf8Error),
+
+    #[error("unexpected tag: found {0}, expected {1}")]
+    UnexpectedTag(String, String),
+
+    #[error("unexpected error: {0}, expected tag {1}")]
+    UnexpectedError(anyhow::Error, String),
+
+    #[error("unexpected event: expected tag {0}")]
+    UnexpectedEvent(String),
+
+    #[error("unexpected EOF: expected tag {0}")]
+    UnexpectedEOF(String),
+
+    #[error("tag not found: {0}")]
+    TagNotFound(String),
+
+    #[error("generic error: {0}")]
+    Generic(String),
+}
+
+#[derive(ThisError, Debug)]
+pub enum EncodingError {
+    #[error("malformed UTF-8: {0}")]
+    Utf8Error(#[from] FromUtf8Error),
+
+    /// Error while parsing (malformed?) XML.
+    #[error("XML error: {0}")]
+    XmlError(#[from] XmlError),
+}
+
+pub type Result<T> = result::Result<T, Error>;
```

Comment:
This is convention. Most libraries seem to export an Error type and a Result type.

Reply:
Yeah, I guess my qualm is less with your code and more with the convention that's formed.  Feel free to disregard this comment.  Read on if you're curious about why I dislike the convention, lol.

Since Error and Result are names that are so common (and used by the standard library), it can get confusing if a library suddenly starts shadowing these names and the dev didn't expect it (e.g. you 'use *' for a particular crate you make heavy use of, and suddenly Result and Error are coming from that crate instead of the standard library).